pubsub-mapped-api/src/main/java/com/google/pubsub/clients/consumer/KafkaConsumer.java

```diff
@@ -0,0 +1,592 @@
+    throw new NotImplementedException("Not yet implemented");
+  }
+
+  public Set<String> subscription() {
+    return subscriptions.keySet();
+  }
+
+  /*
+  For every new consumer, a temporary subscription is created. After "unsubscribe" operation
+  temporary subscriptions are removed. After an error while creating subscriptions the attempt
+  is made to delete all temporary subscriptions made in this call to "subscribe".
+
+  Every version of subscribe is replacing old subscriptions with new ones.
+
+  Temporary subscription names are built as a combination of timestamp and randomly generated UUID.
+   */
+  public void subscribe(Collection<String> topics, ConsumerRebalanceListener listener) {
+    checkSubscribePreconditions(topics);
+
+    unsubscribe();
+    List<ResponseData<Subscription>> futureSubscriptions = deputePubsubSubscribes(topics);
+    Map<String, Subscription> subscriptionMap = getNewSubscriptionMap(futureSubscriptions);
+
+    log.debug("Subscribed to topic(s): {}", Utils.join(subscriptionMap.keySet(), ", "));
+    subscriptions = ImmutableMap.copyOf(subscriptionMap);
+  }
+
+  private List<ResponseData<Subscription>> deputePubsubSubscribes(Collection<String> topics) {
+    Timestamp timestamp = new Timestamp(System.currentTimeMillis());
+
+    List<ResponseData<Subscription>> responseDatas = new ArrayList<>();
+    Set<String> usedNames = new HashSet<>();
+
+    for(String topic: topics) {
+      if(!usedNames.contains(topic)) {
+        String subscriptionString = SUBSCRIPTION_PREFIX + topic + "_" + timestamp.getTime()
+            + "_" + UUID.randomUUID();
+        ListenableFuture<Subscription> deputedsubscription =
+            deputeSinglePubsubSubscription(subscriptionString, topic);
+
+        responseDatas.add(new ResponseData<>(topic, subscriptionString, deputedsubscription));
+        usedNames.add(topic);
+      }
+    }
+    return responseDatas;
+  }
+
+  private ListenableFuture<Subscription> deputeSinglePubsubSubscription(String subscriptionString,
+      String topicName) {
+    SubscriberFutureStub subscriberFutureStub = stubCreator.getSubscriberFutureStub();
+    return subscriberFutureStub
+        .createSubscription(Subscription.newBuilder()
+            .setName(subscriptionString)
+            .setTopic(TOPIC_PREFIX + topicName)
+            .build());
+  }
+
+  private Map<String, Subscription> getNewSubscriptionMap(
+      List<ResponseData<Subscription>> responseDatas) {
+    Map<String, Subscription> subscriptionMap = new HashMap<>();
+
+    for(ResponseData<Subscription> responseData: responseDatas) {
+      try {
+        Subscription s = responseData.getRequestListenableFuture().get();
+        subscriptionMap.put(responseData.getTopicName(), s);
+      } catch (InterruptedException e) {
+        //if an error is thrown, attempt to delete subscriptions created in this loop
+        unsubscribe(subscriptionMap.values());
+        throw new InterruptException(e);
```

Comment:
this is incorrect.  unsubscribe should happen in a finally block, not like this.